images/istio/tests/main.tf

```diff
@@ -29,19 +33,126 @@ data "oci_exec_test" "operator-version" {
   script = "docker run --rm $IMAGE_NAME version"
 }
 
-data "oci_string" "ref" { input = var.digests.operator }
+data "oci_string" "operator-ref" { input = var.digests.operator }
+data "oci_string" "proxy-ref" { input = var.digests.proxy }
 
 resource "helm_release" "operator" {
-  name = "operator"
-
-  # There's no official helm chart for the istio operator
+  name             = "operator"
+  namespace        = local.namespace
+  create_namespace = true
+  # there's no official helm chart for the istio operator
   repository = "https://stevehipwell.github.io/helm-charts/"
   chart      = "istio-operator"
 
   values = [jsonencode({
     image = {
-      repository = data.oci_string.ref.registry_repo
-      tag        = data.oci_string.ref.pseudo_tag
+      repository = data.oci_string.operator-ref.registry_repo
+      tag        = data.oci_string.operator-ref.pseudo_tag
+    }
+  })]
+}
+
+resource "random_pet" "suffix" {}
+
+locals {
+  namespace = "istio-system-${random_pet.suffix.id}"
+}
+resource "helm_release" "base" {
+  name             = "${local.namespace}-base"
+  namespace        = local.namespace
+  create_namespace = true
+  repository       = "https://istio-release.storage.googleapis.com/charts/"
+  chart            = "base"
+  replace          = true # Allow reinstallation - as CRDs are not reinstalled anyway.
+  values = [jsonencode({
+    global = {
+      istioNamespace = local.namespace
+    }
+  })]
+}
+
+resource "helm_release" "istiod" {
+  depends_on       = [helm_release.base]
+  name             = "${local.namespace}-istiod"
+  namespace        = local.namespace
+  create_namespace = true
+  repository       = "https://istio-release.storage.googleapis.com/charts/"
+  chart            = "istiod"
+  values = [jsonencode({
+    # Set the revision so that only namespace with istio.io/rev=local.namespace
+    # will be managed.
+    revision = local.namespace
+    pilot = {
+      image = var.digests.pilot
+    }
+    global = {
+      istioNamespace = local.namespace
+      # We have to trim the suffix and specify it in `image`, because this
+      # Helm chart does not like slashes in the image name.
+      hub = replace(data.oci_string.proxy-ref.registry_repo, "/istio-proxy", "")
```

Comment:
🤔 I'm puzzled at how this works

Reply:
If the proxy image is `gcr.io/test/istio/image/istio-proxy:unused@sha256:deadbeef`, the chart expects:
* `hub` = gcr.io/test/istio/image
* `image` = istio-proxy
* `tag` = `unused@...`

The `oci_string.foo.registry_repo` would give me gcr.io/test/istio/image/istio-proxy, hence I need to chop that last part off.